go/vt/vtgate/vindexes/cfc.go

```diff
@@ -0,0 +1,326 @@
+	out := make([]bool, len(ids))
+	for i := range ids {
+		v, err := vind.computeKsid(ids[i].ToBytes(), false)
+		if err != nil {
+			return nil, err
+		}
+		out[i] = bytes.Equal(v, ksids[i])
+	}
+	return out, nil
+}
+
+// Map can map ids to key.Destination objects.
+func (vind *CFC) Map(cursor VCursor, ids []sqltypes.Value) ([]key.Destination, error) {
+	out := make([]key.Destination, len(ids))
+	for i, id := range ids {
+		v, err := vind.computeKsid(id.ToBytes(), false)
+		if err != nil {
+			return nil, err
+		}
+		out[i] = key.DestinationKeyspaceID(v)
+	}
+	return out, nil
+}
+
+// PrefixVindex switches the vindex to prefix mode
+func (vind *CFC) PrefixVindex() SingleColumn {
+	return vind.prefixVindex
+}
+
+func addOne(value []byte) []byte {
+	n := len(value)
+	overflow := true
+	for i := n - 1; i >= 0; i-- {
+		if value[i] < 255 {
+			value[i]++
+			overflow = false
+			break
+		} else {
+			value[i] = 0
+		}
+	}
+	if overflow {
+		return nil
+	}
+	return value
+}
+
+type prefixCFC struct {
+	name string
+	cfc  *CFC
+}
+
+func (vind *prefixCFC) String() string {
+	return vind.name
+}
+
+func (vind *prefixCFC) Cost() int {
+	return 1
+}
+
+func (vind *prefixCFC) IsUnique() bool {
+	return false
+}
+
+func (vind *prefixCFC) NeedsVCursor() bool {
+	return false
+}
+
+// Verify returns true if ids maps to ksids.
+func (vind *prefixCFC) Verify(vc VCursor, ids []sqltypes.Value, ksids [][]byte) ([]bool, error) {
+	// prefixCFC is only active for 'LIKE' expr. Verify is used by 'INSERT' so we re-use CFC.Verify()
+	return vind.cfc.Verify(vc, ids, ksids)
+}
+
+// Map can map ids to key.Destination objects.
+func (vind *prefixCFC) Map(cursor VCursor, ids []sqltypes.Value) ([]key.Destination, error) {
+	out := make([]key.Destination, len(ids))
+	for i, id := range ids {
+		value := id.ToBytes()
+		// we only consider the string up to the first whild char as the prefix.
+		p := bytes.IndexByte(value, '%')
+		if p < 0 {
+			// if there is no wild char, we use the whole string
+			p = len(value)
+		}
+		begin, err := vind.cfc.computeKsid(value[:p], true)
+		if err != nil {
+			return nil, err
+		} else if len(begin) == 0 {
+			// prefix is too short, this is typically the case when the prefix
+			// is smaller than the first offset when using hash.
+			out[i] = key.DestinationAllShards{}
+		} else {
+			// the prefix maps to keyspace id corresponding to value and plus one.
+			// that is [ keyspace_id, keyspace_id + 1 ).
+			end := make([]byte, len(begin))
+			copy(end, begin)
+			end = addOne(end)
+			out[i] = key.DestinationKeyRange{
+				KeyRange: &topodatapb.KeyRange{
+					Start: begin,
+					End:   end,
+				},
+			}
+		}
+
+	}
+	return out, nil
+}
+
+// we don't use the full hashed value because it's very long.
+// keyrange resolution is done via comparing []byte so longer
+// keyspace ids have performance impact.
+func md5hash(in []byte) []byte {
+	n := len(in)
+	out := vMD5Hash(in)
+	if n < len(out) {
```

Comment:
Based on your current implementation, this means that your unhashed data is coupled to the hashed data size. I cannot choose to take an 8-byte value, and hash it to a 4-byte value.

Unless you do all the hashing yourself, and leave `hash` and `offsets` defined, which I think is perfectly fine.

I'm more pointing out that, I think this is going to limit the utility of the hashing.

If I take my 8-byte value, choose 4 of the bytes, and then hash that, the composite hash might not be distributed as well, if I'm wrong about the distribution of those 4 bytes.

Though, the approach you're taking now has similar issues. If the hashing function returns an N-byte array, and you only take the first M-bytes of that array, it might not be distributed very well. If you really want an M-byte array instead, you should choose a function that is designed to return M-bytes, while still acting as a good hash function.

All of this makes me question whether doing automatic hashing is worth it. Or if the client should compute their own hashes, that way they can make their own determination how good of a hash they need.

The other limiting requirement here is the fact that the inputs must be fixed-width. With the client defining the function that produces the column value, they can support variable-width component functions, e.g. `identity(c1) concat identity(c2)`, or `reverse(c1) concat reverse(c2)`. You don't have that kind of flexibility with this hashing system. Which isn't necessarily a bad thing. But again, I'm wondering how much these hashes would end up being used, versus manual value mapping in the client.

Reply:
I think it would help to see some real-world examples of how you are planning on using this. The RFC and tests only contain toy schemas. What would it look like in a production schema, which wanted to achieve soft-colocation using this CFC concept.

Reply:
Thanks for the comments. I debated with myself for whether to support hashing at all, very similar to what you said. Finally I decided to add it as a way of convenience so that application doesn't have to do it. At the end of the day, there are only a limited number of shards, regardless how we hash it it's just a best effort to distribute the keyspace ids. When the components in a composite index have some narrow value range, e.g. some column in the composite index may be an enum, adding a parameter to vschema is arguably more convenient than changing application code to produce a hashed concatenated column.

Reply:
A more concrete example:  inventory system. A simplified data model would be that, there is 'product' within which there are many items. The mental model is, there could be couple orders of magnitude more items than products. Items check-in and check-out the inventory for sales/return/restock etc. We want to have per product view of its inventory status, e.g. how many items in stock within an hour block, average time in the warehouse in the last 7 days, return rate, etc.
```
product {
  product_id
  product_name
  msrp
}

item {
  item_id
  product_id
  serial_no
  price
  check_in_at
  check_out_at
  status
}
```

Let's assume the status change or item creation has very high throughput. We can shard by item id and create a global secondary index on product id and normalize all the columns needed in the query with the index. But we have to perform two inserts to insert one item, one in the item table the other in the global secondary index table on another shard. It has its pros and cons. CFC vindex provides another option. We can have the item table look like
```
item {
  item_id
  product_id
  serial_no
  price
  check_in_at
  check_out_at
  status
  keyid // first 2 bytes of product_id + item_id
}
```
If we define CFC vindex on `keyid`, given a product_id, we can then use 
```
select count(*) from items where keyid like 'xx%' and product_id = 'xxxxxx' and check_in_at > t0 and check_in_at <= t1 and check_out_at is null
```
to find out within (t0, t1) how many items for the product whose id is 'xxxxxx'.  This query only hit a subset of shards instead of all shards. (i.e. 'xx123' and 'xx456' may very well belong to two different shards) The write throughput of inserting/updating an item is spread across more than one shard so that its qps is not bottlenecked by a single shard.